### PR TITLE
[AArch64][GlobalISel] Combine G_CONCAT_VECTORS with Illegal Form

### PR DESCRIPTION
**Pull request by chuongg3**

### CODE REVIEW

File: llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h [hunk 1]
```diff
@@ -359,6 +359,108 @@ class LegalizationArtifactCombiner {
...
+    for (unsigned i = 0; i < NumSrc; i++) {
+      Register CurrReg = MI.getSourceReg(i);
+      LLT CurrTy = MRI.getType(CurrReg);
+
+      // Only handle cases where all source types are the same type
+      if (CurrTy.getSizeInBits() != SrcTy.getSizeInBits())
+        return false;
+
+      // Register defined with G_IMPLIT_DEF
+      if (isImpDefVRegValWithLookThrough(CurrReg, MRI)) {
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h:409]: It shouldn't be required to special case implicit_defs

File: llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h [hunk 2]
```diff
@@ -359,6 +359,108 @@ class LegalizationArtifactCombiner {
...
+        NumUndef++;
+        SrcRegs.push_back(CurrReg);
+      }
+
+      else if (MachineInstr *DefMI = getDefIgnoringCopies(CurrReg, MRI);
+               DefMI->getOpcode() == TargetOpcode::G_BITCAST) {
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h:415]: bitcast isn't defined as a legalization artifact, so the artifact combiner shouldn't be specially handling them

Comment (davemgreen) [reply llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h:415]: Should we be handling bitcast as legalization artifacts? We need a way to clean away illegal nodes between legal ones as we are legalizing.

**Review by aemerson — COMMENTED** [review]
This feels like something that should be automatically simplified away by the user of the concat, using the ArtifactValueFinder. We should be able to look through the concat of `{source_val, undef, undef, undef}` and directly forward the `source_val` register to the user.

The idea is that we as much as possible avoid adding fragile combines/transformations to the artifact combiner to keep it stable and resilient to changing IR inputs.

File: llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h (hunk 2 above)
Comment (arsenm) [reply llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h:415]: I think the answer is no, as we can represent bitcast with the merge/unmerge primitives. But maybe?

File: llvm/lib/CodeGen/GlobalISel/Utils.cpp [hunk 3]
```diff
@@ -411,6 +411,44 @@ std::optional<APInt> getCImmOrFPImmAsAPInt(const MachineInstr *MI) {
 
 } // end anonymous namespace
 
+// Checks if a register has been defined by G_IMPLICIT_DEF previously
+bool llvm::isImpDefVRegValWithLookThrough(Register VReg,
+                                          const MachineRegisterInfo &MRI) {
```
Comment (aemerson) [inline llvm/lib/CodeGen/GlobalISel/Utils.cpp:416]: I'd just move this to the legalizer since it's not going to be used by anything else.

**Review by aemerson — COMMENTED** [review]
I think this is a very specific transform to get around our `<4 x s8>` issue but unfortunately I don't have a better suggestion.

**Review by aemerson — APPROVED** [review]

**Comment by arsenm** [comment]
> This feels like something that should be automatically simplified away by the user of the concat, using the ArtifactValueFinder. We should be able to look through the concat of `{source_val, undef, undef, undef}` and directly forward the `source_val` register to the user.
> 
> The idea is that we as much as possible avoid adding fragile combines/transformations to the artifact combiner to keep it stable and resilient to changing IR inputs.

Yes, but this is just adding another one? I really don't think the artifact combiner has any business checking for whether a source is implicit_def. Implicit_def isn't structurally significant. We also should be moving all of this logic to use the ArtifactValueFinder

**Comment by chuongg3** [comment]
Issue fixed by #96492

**Pull request closed without merging**